Issue/webpack 5 upgrade

cayb0rg opened the pull request:
## MWDK Upgrade
* Upgrades all of the packages in the Materia Widget Development Kit to their latest versions and removes deprecated packages. 
* Server is now created in express.js and uses ``webpack-dev-middleware`` to compile and view assets.
* Adds a loading screen for installing widgets to local Materia
* Adds a scorescreen page. Custom scorescreens are supported if the author uploads sample score data from Materia.

## Requirements: 
- Node v16 or later (including v18)
- Yarn

## Local setup
<del>
### Materia
The current Materia widget-dependencies-package does not support the MWDK. UPDATE: the MWD package has been updated. **These steps are no longer needed**.
1. Checkout this [react/widget-dependencies-package-part-three branch](https://github.com/cayb0rg/Materia/tree/react/widget-dependencies-package-part-three) of Materia. This is required to load the player and creator.
2. Install Materia if you have not.
3. Run ``yarn build`` inside Materia's root folder. 
4. Run ``yarn link`` inside Materia/public/dist.
5. Start Materia (run ``docker-compose up`` in Materia/docker and ``yarn dev`` in Materia/src)
</del>

### Widget
Each widget may need a few small changes to support Webpack v5. Here is an incomprehensive list of the possible changes:

1. New webpack.config.js entry rules:
- Include only entries that have HTML files (except for markdown files), as these will be used by HtmlWebpackPlugin. Include the HTML file as the first element, then include any necessary JS and CSS in the array.
- There should be no extension in the entry's name.
- CSS and Markdown guides should not have their own entries

It should look similar to:
```
const entries = {
	'player': [
		path.join(srcPath, 'player.html'),
		path.join(srcPath, 'player.js'),
		path.join(srcPath, 'player.scss')
	],
	'creator': [
		path.join(srcPath, 'creator.html'),
		path.join(srcPath, 'creator.js'),
		path.join(srcPath, 'creator.scss'),
	],
	'scoreScreen': [
		path.join(srcPath, 'scoreScreen.html'),
		path.join(srcPath, 'scoreScreen.js'),
		path.join(srcPath, 'scoreScreen.scss'),
	]
}
```

2. Either include CSS inside your JS files for MiniCssExtractPlugin to recognize them, or include them as elements in the entry arrays (see above).
3. Use absolute urls for assets like loading fonts
<img width="586" alt="image" src="https://github.com/ucfopen/Materia-Widget-Dev-Kit/assets/46247315/61ceaa96-a7e7-40ce-a1e7-31f212f6fe00">

4. Remove the following rules from ``moduleRules``, if applicable: ``rules.loadAndCompileMarkdown`` and``rules.loadAndPrefixCSS``
5. Update the NPM scripts inside ``package.json`` to:
```
"start": "mwdk-start",
"build": "mwdk-build-prod",
"build-dev": "mwdk-build-dev"
```
- mwdk-start: This starts the Express server and sets the webpack configuration to development mode.
- mwdk-build-prod: This runs webpack in production mode. See the output in /build folder.
- mwdk-build-dev: This runs webpack in development mode. See the output in /build folder.

Look at an [example](https://github.com/ucfopen/syntax-sorter-materia-widget/compare/master...cayb0rg:issue/webpack-5-upgrade) for the changes that need to be made (ignore the images).

#### AngularJS Widgets
AngularJS widgets may encounter an angular injection error. This usually only occurs with widgets that have directives, services, and controllers loaded in separate files. There are two ways to fix this. One way is to do what This or That does, and set up the directives, controllers, and services in a single file (for example, see [creator.js](https://github.com/ucfopen/this-or-that-materia-widget/blob/master/src/creator.js). The other, more difficult way I've found is to manually bootstrap the angular modules. 
1. Remove ``ng-app`` from body or HTML tags.
2. Move the main scripts such as ``player.js`` and ``creator.js`` to the end of the body tag
3. Inside the files declaring the angular modules, remove the variable declaration when initially declaring the module. For example, the matching widget would look like:
![image](https://github.com/ucfopen/Materia-Widget-Dev-Kit/assets/46247315/d10d1b62-f412-4add-9252-74fffe61b467)
4. At the very end of the last file in the respective webpack entry, manually start the module with the command: ``angular.bootstrap(document, ['your-module-name-here'])``. The example above would look like: 
![image](https://github.com/ucfopen/Materia-Widget-Dev-Kit/assets/46247315/7b3413e6-70db-4ac0-b905-b4eaf5731e31)

Look at an [example](https://github.com/ucfopen/matching-materia-widget/compare/master...cayb0rg:matching-materia-widget:mwdk-upgrade) for the changes that need to be made.

### Running this MWDK PR
1. Checkout this PR. 
2. Run ``yarn install && yarn build && yarn link``inside the Materia Widget Development Kit. You may also use ``bun install``.
3. Run ``rm -rf node_modules`` inside your widget. 
4. Enter the git url for MWDK dependency: `"materia-widget-development-kit": "https://github.com/cayb0rg/Materia-Widget-Dev-Kit.git#issue/webpack-5-upgrade"`
7. Run ``yarn install`` or ``bun install``
8. (Optional) Run ``yarn link materia-widget-development-kit``. This will allow you to make changes to the MWDK without having to reinstall it in your widget.
9. Run ``yarn start`` or ``bun start``

## Known Bugs / To-Do
- [x] ~~Angular widgets do not work.~~ (see above)
- [x] ~~Not updating the title of a widget may cause an error when trying to preview it.~~
- [x] ~~Question importer does not work.~~ (fixed)
- [x] ~~MWDK does not run independently of widgets.~~ (fixed)
- [x] URLs for assets could be cleaned up and made clearer.
- [x] ~~Media importer creates incorrect urls in creator~~ (client issue)
- [x] Custom scorescreen support
  - Attempting to call each widget's score_module class functions may be more hassle than it's worth
  - Alternative solution: have the user upload a .json file of sample score data (pulled from the widget_instance_play_scores_get in Materia)
- [x]  ~~Issues resolving babel-loader for React widgets~~ (client issue)
- [x] ~~Getting a Compilation.assets deprecation error in console~~ (fixed)
  - ~~Update ``webpack-generate-widget-hash`` to use [compiler.hooks.compilation.tap](https://webpack.js.org/api/compilation-hooks/#processassets)~~
- [x]  ~~Screenshot annotation tool~~ (fixed)
- [x] Trying to load localhost:8118 before webpack has finished building will not load the demo instance
- [ ] Update all of the widgets' MWDK path
- [x] ~~Icomoon fonts get a 404 once installed in Materia~~

### Updated Widgets (PRs made)
- [x] Adventure
  - Will ship with inventory system
- [x] Crossword (https://github.com/ucfopen/crossword-materia-widget/pull/56)
- [x] Enigma (https://github.com/ucfopen/enigma-materia-widget/pull/45)
- [ ] Equation Sandbox
- [x] Flash Cards (https://github.com/ucfopen/flash-cards-materia-widget/pull/32)
- [ ] Guess the Phrase
  - On hold
- [x] Labeling (https://github.com/ucfopen/labeling-materia-widget/pull/36)
- [x] Last Chance Cadet (https://github.com/ucfopen/last-chance-cadet-materia-widget/pull/39)
- [x] Matching (https://github.com/ucfopen/matching-materia-widget/pull/42)
- [x] Node Graph (https://github.com/ucfcdl/node-graph-widget/pull/19)
- [x] Normal Distribution Calculator (https://github.com/ucfopen/normal-distribution-calculator-materia-widget/pull/37)
- [x] Radar Grapher (https://github.com/ucfopen/radar-grapher-materia-widget/pull/34)
- [x] Secret Spreadsheet (https://github.com/ucfopen/secret-spreadsheet-materia-widget/pull/74)
- [ ] Sequencer
- [ ] Simple Survey
- [x] Slope Finder (https://github.com/ucfopen/slope-finder-materia-widget/pull/32)
- [x] Sort it Out (https://github.com/ucfopen/sort-it-out-materia-widget/pull/57)
- [x] Syntax Sorter (https://github.com/ucfopen/syntax-sorter-materia-widget/pull/63)
- [x] This or That (https://github.com/ucfopen/this-or-that-materia-widget/pull/57)
- [ ] Word Guess
  - On hold
- [ ] Word Search
  - On hold

clpetersonucf commented on qsets/.gitkeep:
At least as of right now, `yarn start` initializes but the webpack dev server errors out as soon as you load `localhost:8118` because the `qsets/` dir is missing.

clpetersonucf commented on express.js:
```diff
@@ -269,372 +336,877 @@ var resizeImage = (size, double) => {
+	// 2. filter for materia-web image and named xxxx_phpfpm_1 name
+	// 3. pick the first line
+	// 4. pick the container name
+	let targetImage = execSync('docker ps -a --format "{{.Image}} {{.Names}}" | grep -e ".*materia:.* docker_app_.*" | head -n 1 | cut -d" " -f2');
```
This line continues to give us problems. Some versions of docker-compose use hyphens, others use underscores, and the server does not handle a search failure well. We can modify the search pattern grep is using to: `".*materia:.* docker[-_]app[-_].*"` which will catch either option.

clpetersonucf replied:
Also it'd be nice if the server gave a more well-handled response if the docker install process runs into an issue

clpetersonucf commented on .vscode/settings.json:
Is this file (and corresponding directory) required?

clpetersonucf commented on package.json:
```diff
@@ -7,52 +7,78 @@
+    "hbs": "^4.2.0",
+    "html-loader": "^4.2.0",
+    "html-webpack-plugin": "^5.5.3",
+    "materia-widget-dependencies": "0.0.5",
```
MWD is now at `0.2.0` and should include the missing `path.js` file.

clpetersonucf commented on webpack-widget.js:
```diff
@@ -146,7 +145,7 @@ const getDefaultCopyList = () => {
 		console.log(`===== USING ${devDemo} ====`)
 		copyList.push({
-			flatten: true,
+			// flatten: true,
```
Required?

clpetersonucf reviewed (changes requested):
Really great work here so far. With limited issues I was able to get one of your widget PRs running locally and installed to dev Materia.

Some high-level comments:

- As commented, MWD has been updated to hopefully (?) include the most recent/missing dependencies. It's worth confirming 0.2.0 works out of the box, and replace the reference to it in the package.json when possible.
- I _love_ the little progress bar when installing to local Materia. That's a nice touch.
- I haven't tested the score screen feature, but the instructions could be a little clearer as to what's required. If the score screen options aren't ready for prime-time, we could also potentially disable them until they have more time in the oven.
- Previously the MWDK would allow you to save draft/publish a qset to test in the player. Right now it looks like the demo is the only qset that works with the player. Is that the case? Are there significant challenges to re-implementing cached qsets?
- Would it be possible to render Player and Creator guides, if they exist, within the MWDK? It looks like we already have view templates to support them.

Otherwise, just a little bit of polish based on feedback items, and making sure all current changes on this branch are final, required, and complete, and it looks like it might be good to go.

cayb0rg commented:
Thanks for the feedback! 

New changes:
- Updated materia-widget-dependencies package version to 0.2.0
- Fixed link to `guides.js` in guide pages
- Updated scorescreen instructions
- The kit should support `demo` and one additional instance. I've added the ability to have multiple widget instances.
- Made the widget installation process more verbose so that all errors are forwarded to the user as they are received

clpetersonucf reviewed (approved):
There might be some edge case issues we haven't encountered yet, but nothing show-stopping. By all accounts the new dev kit does the job required and works great.

Really excellent work. This was a significant undertaking and you knocked it out of the park.

FrenjaminBanklin commented on express.js:
```diff
@@ -1,42 +1,76 @@
-
-var webPackMiddleware = false;
-var hasCompiled = false;
+const path            	= require('path');
+const fs              	= require('fs');
+const express         	= require('express')
+const qsets           	= path.join(__dirname, 'qsets');
+const yaml            	= require('yamljs');
+const { execSync }    	= require('child_process');
+const waitUntil       	= require('wait-until-promise').default
+const { v4: uuidv4 }    = require('uuid')
+const sharp           	= require('sharp')
+const util				= require('util');
+const cors 				= require('cors')
+const hbs				= require('hbs');
+
+// common paths used here
+const srcPath 				= path.join(process.cwd(), 'src') + path.sep
+const outputPath 			= path.join(process.cwd(), 'build') + path.sep
+
+// Webpack middleware setup
+const webpack 							= require('webpack');
+const webpackDevMiddleware 	= require('webpack-dev-middleware');
```
This doesn't really matter, but stick to spaces for inline alignment.

Would prefer to remove all the extraneous white space, personally.

FrenjaminBanklin reviewed (approved):
Super minor nitpick with some inline spacing, but nothing I'd hold up a release for.

I couldn't get the score screen working with valid JSON, but considering the latest version of the MWDK doesn't have a functional score screen either I'm not sure I'd hold up a release for that, either.

Everything looks like it's working well otherwise.
